test/qa/lib/firewall.py

```diff
@@ -93,98 +96,195 @@
+            f"-A OUTPUT -o {interface} -m connmark --mark 0xe1f1 -m comment --comment nordvpn -j ACCEPT",
+            f"-A OUTPUT -o {interface} -m comment --comment nordvpn -j DROP"
+        ]
+    elif len(ports) > 2:
+        ports.reverse()
+
+        result = []
+
+        for port in ports:
+                result.extend([
+                    f"-A INPUT -i {interface} -p {protocol} -m {protocol} --dport {port} -m comment --comment nordvpn -j ACCEPT",
+                    f"-A INPUT -i {interface} -p {protocol} -m {protocol} --sport {port} -m comment --comment nordvpn -j ACCEPT"
+                ])
+        result += \
+        [
+            f"-A INPUT -i {interface} -m connmark --mark 0xe1f1 -m comment --comment nordvpn -j ACCEPT",
+            f"-A INPUT -i {interface} -m comment --comment nordvpn -j DROP",
+        ]
+
+        for port in ports:
+            result.extend([
+                f"-A OUTPUT -o {interface} -p {protocol} -m {protocol} --dport {port} -m comment --comment nordvpn -j ACCEPT",
+                f"-A OUTPUT -o {interface} -p {protocol} -m {protocol} --sport {port} -m comment --comment nordvpn -j ACCEPT"
+            ])
+
+        result += \
+        [
+            f"-A OUTPUT -o {interface} -m mark --mark 0xe1f1 -m comment --comment nordvpn -j CONNMARK --save-mark --nfmask 0xffffffff --ctmask 0xffffffff",
+            f"-A OUTPUT -o {interface} -m connmark --mark 0xe1f1 -m comment --comment nordvpn -j ACCEPT",
+            f"-A OUTPUT -o {interface} -m comment --comment nordvpn -j DROP"
+        ]
+        return result
+
+
+def is_active(ports: list[str]="", protocol="", subnets: list[str]="") -> bool:
+    """ returns True when all expected rules are found in iptables in matching order """
+    print(sh.ip.route())
+
+    default_route_interface = sh.ip.route.show("default").split(None)[4]
 
     try:
-        # Get VPN server's IP address
-        status = sh.grep(sh.nordvpn.status(), "IP")
-        _, server_ip = status.split(None, 2)
-    except sh.ErrorReturnCode:
+        server_ip = settings.get_server_ip()
+    except IndexError:
         server_ip = ""
-    print("Default gateway:", iface, "Server's IP:  ", server_ip)
 
-    rules = _get_firewall_rules(
-        daemon.is_killswitch_on(), server_ip, iface, port, protocol, subnet
+    print("Default gateway:", default_route_interface, "; Server IP:", server_ip)
+
+    expected_rules = _get_firewall_rules(
+        daemon.is_killswitch_on(), server_ip, default_route_interface, ports, protocol, subnets
     )
-    print("Expected rules:\n", rules)
+    print("Expected rules:\n")
+    for rule in expected_rules:
+        print(rule)
+    print("\n")
 
     current_rules = _get_iptables_rules()
-    print("Current rules:\n", current_rules)
+    print("Current rules:\n")
+    for rule in current_rules:
+        print(rule)
```

Comment:
shouldn't each rule be printed on a new line?

Reply:
Checked this once again, it seems like we are indeed printing elements to new lines :smile: